packages/connect-react/src/hooks/use-app.tsx

```diff
@@ -23,3 +26,64 @@ export const useApp = (slug: string, opts?:{ useQueryOpts?: Omit<UseQueryOptions
     app: query.data?.data,
   };
 };
+
+type AppResponseWithExtractedCustomFields = AppResponse & {
+  extracted_custom_fields_names: string[]
+}
+
+type AppCustomField = {
+  name: string
+  optional?: boolean
+}
+
+type OauthAppPropValue = PropValue<"app"> & {
+  oauth_access_token?: string
+}
+
+function getCustomFields(app: AppResponse): AppCustomField[] {
+  const isOauth = app.auth_type === "oauth"
+  const userDefinedCustomFields = JSON.parse(app.custom_fields_json || "[]")
+  if ("extracted_custom_fields_names" in app && app.extracted_custom_fields_names) {
+    const extractedCustomFields = ((app as AppResponseWithExtractedCustomFields).extracted_custom_fields_names || []).map(
+      (name) => ({
+        name,
+      }),
+    )
+    userDefinedCustomFields.push(...extractedCustomFields)
+  }
+  return userDefinedCustomFields.map((cf: AppCustomField) => {
+    return {
+      ...cf,
+      // if oauth, treat all as optional (they are usually needed for getting access token)
+      optional: cf.optional || isOauth,
+    }
+  })
+}
+
+export function appPropError(prop: ConfigurablePropApp, value: unknown, app: AppResponse | undefined): string | undefined {
```

Comment:
Does `prop` need to be here if it's not being used?

There are a couple of parameters being passed in. Does it make sense to just use an object?